File: src/main/java/kg/manasuniversity/usbtypec/manashelper/user/service/obis/parser/ObisParser.java
package kg.manasuniversity.usbtypec.manashelper.user.service.obis.parser;

import kg.manasuniversity.usbtypec.manashelper.user.exception.ObisPageParserException;
import kg.manasuniversity.usbtypec.manashelper.user.model.Exam;
import kg.manasuniversity.usbtypec.manashelper.user.model.LessonAttendance;
import kg.manasuniversity.usbtypec.manashelper.user.model.LessonExams;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class ObisParser {

    public List<LessonExams> parseAttendanceHtmlPage(String html) throws ObisPageParserException {
        Document doc = Jsoup.parse(html);

        Elements tableBodies = doc.select("tbody");
        if (tableBodies.isEmpty()) {
            throw new ObisPageParserException("На странице в OBIS отсутствует таблица с оценками.");
        }

        Element tbody = tableBodies.last();
        if (tbody == null) {
            throw new ObisPageParserException("На странице в OBIS отсутствует таблица с оценками.");
        }

        Elements rows = tbody.select("> tr");

        List<LessonExams> lessons = new ArrayList<>();
        int i = 0;
        while (i < rows.size()) {
            Element mainRow = rows.get(i);
            Elements tds = mainRow.select("> td");

            if (tds.size() == 5) {
                String lessonCode = textOrNull(tds.get(1));
                String lessonName = textOrNull(tds.get(2));

                int rowspan = 1;
                if (tds.get(0).hasAttr("rowspan")) {
                    rowspan = Integer.parseInt(tds.get(0).attr("rowspan"));
                }

                List<Exam> exams = new ArrayList<>();

                exams.add(new Exam(
                    textOrNull(tds.get(3)),
                    textOrNull(tds.get(4))
                ));

                for (int j = 1; j < rowspan && i + j < rows.size(); j++) {
                    Element nextRow = rows.get(i + j);
                    Elements nextTds = nextRow.select("> td");

                    if (nextTds.size() == 2) {
                        exams.add(new Exam(
                            textOrNull(nextTds.get(0)),
                            textOrNull(nextTds.get(1))
                        ));
                    }
                }

                lessons.add(new LessonExams(
                    lessonName,
                    lessonCode,
                    exams
                ));

                i += rowspan;
            } else {
                i++;
            }
        }

        return lessons;
    }

    public List<LessonAttendance> parseLessonsAttendancePage(String html) throws ObisPageParserException {
        Document doc = Jsoup.parse(html);

        Element table = doc.selectFirst("table");
        if (table == null) {
            throw new ObisPageParserException("На странице в OBIS отсутствует таблица с посещаемостью.");
        }

        Elements rows = table.select("tr");
        List<LessonAttendance> lessons = new ArrayList<>();

        for (int i = 1; i < rows.size(); i++) { // skip header
            Element row = rows.get(i);
            Elements tds = row.select("td");

            if (tds.size() != 9) {
                continue;
            }

            String lessonCode = tds.get(1).text().trim();
            String lessonName = tds.get(2).text().trim();

            String theoryPercentage = tds.get(4).text().replace("%", "").trim();
            String practicePercentage = tds.get(6).text().replace("%", "").trim();

            lessons.add(new LessonAttendance(
                lessonName,
                lessonCode,
                tryParseDouble(theoryPercentage),
                tryParseDouble(practicePercentage)
            ));
        }

        return lessons;
    }

    /**
     * Retrieve CSRF token from OBIS login page HTML.
     *
     * @param html HTML content of the login page.
     * @return CSRF token as a String.
     * @throws ObisPageParserException if the CSRF token is not found.
     */
    public String parseLoginPageCsrfToken(String html) throws ObisPageParserException {
        Document doc = Jsoup.parse(html);
        Element csrfInput = doc.selectFirst("form input[name=_csrf]");
        if (csrfInput == null) {
            throw new ObisPageParserException("CSRF токен не найден на странице логина OBIS.");
        }
        String csrfToken = csrfInput.attr("value");
        if (csrfToken.isBlank()) {
            throw new ObisPageParserException("CSRF токен не найден на странице логина OBIS.");
        }
        return csrfToken;
    }

    /**
     * Get trimmed text from an Element or return null if empty.
     *
     * @param element JSoup Element.
     * @return Trimmed text or null.
     */
    private static String textOrNull(Element element) {
        String text = element.text().trim();
        return text.isEmpty() ? null : text;
    }

    /**
     * Try to parse a String to Double, return null if fails.
     *
     * @param value String value to parse.
     * @return Parsed Double or null.
     */
    private static Double tryParseDouble(String value) {
        try {
            return Double.parseDouble(value);
        } catch (Exception e) {
            return null;
        }
    }
}

File: src/manashelper/scraping/obis_parser.py
from dataclasses import dataclass

from bs4 import BeautifulSoup
from bs4.element import Tag


class ObisParseError(Exception):
    pass


@dataclass(frozen=True, slots=True)
class ScrapedExam:
    name: str | None
    score: str | None


@dataclass(frozen=True, slots=True)
class ScrapedLessonExams:
    lesson_name: str | None
    lesson_code: str | None
    exams: list[ScrapedExam]


@dataclass(frozen=True, slots=True)
class ScrapedLessonAttendance:
    lesson_name: str
    lesson_code: str
    theory_skips_percentage: float | None
    practice_skips_percentage: float | None


def parse_login_page_csrf_token(html: str) -> str:
    soup = BeautifulSoup(html, "lxml")
    csrf_input = soup.select_one("form input[name=_csrf]")
    if csrf_input is None:
        raise ObisParseError("CSRF токен не найден на странице логина OBIS.")

    token = csrf_input.get("value")
    if not isinstance(token, str) or not token:
        raise ObisParseError("CSRF токен не найден на странице логина OBIS.")
    return token


def parse_exam_grades_page(html: str) -> list[ScrapedLessonExams]:
    soup = BeautifulSoup(html, "lxml")
    table_bodies = soup.select("tbody")
    if not table_bodies:
        raise ObisParseError("На странице в OBIS отсутствует таблица с оценками.")

    rows = table_bodies[-1].select(":scope > tr")

    lessons: list[ScrapedLessonExams] = []
    i = 0
    while i < len(rows):
        cells = rows[i].select(":scope > td")

        if len(cells) != 5:
            i += 1
            continue

        lesson_code = _text_or_none(cells[1])
        lesson_name = _text_or_none(cells[2])

        rowspan_attr = cells[0].get("rowspan")
        rowspan = int(rowspan_attr) if isinstance(rowspan_attr, str) else 1

        exams = [ScrapedExam(name=_text_or_none(cells[3]), score=_text_or_none(cells[4]))]

        for j in range(1, rowspan):
            if i + j >= len(rows):
                break
            next_cells = rows[i + j].select(":scope > td")
            if len(next_cells) == 2:
                exams.append(ScrapedExam(name=_text_or_none(next_cells[0]), score=_text_or_none(next_cells[1])))

        lessons.append(ScrapedLessonExams(lesson_name=lesson_name, lesson_code=lesson_code, exams=exams))
        i += rowspan

    return lessons


def parse_attendance_page(html: str) -> list[ScrapedLessonAttendance]:
    soup = BeautifulSoup(html, "lxml")
    table = soup.select_one("table")
    if table is None:
        raise ObisParseError("На странице в OBIS отсутствует таблица с посещаемостью.")

    rows = table.select("tr")
    lessons: list[ScrapedLessonAttendance] = []

    for row in rows[1:]:
        cells = row.select("td")
        if len(cells) != 9:
            continue

        lesson_code = cells[1].get_text().strip()
        lesson_name = cells[2].get_text().strip()
        theory_percentage = cells[4].get_text().replace("%", "").strip()
        practice_percentage = cells[6].get_text().replace("%", "").strip()

        lessons.append(
            ScrapedLessonAttendance(
                lesson_name=lesson_name,
                lesson_code=lesson_code,
                theory_skips_percentage=_try_parse_float(theory_percentage),
                practice_skips_percentage=_try_parse_float(practice_percentage),
            )
        )

    return lessons


def _text_or_none(cell: Tag) -> str | None:
    text = cell.get_text().strip()
    return text or None


def _try_parse_float(value: str) -> float | None:
    try:
        return float(value)
    except ValueError:
        return None
